### PR TITLE
Add support for Personal Access Token Authentication

### PR DESCRIPTION
**Pull request by rporres**
https://github.com/andygrunwald/go-jira supports natively Personal Token Authentication (PAT). This PR brings it to jiralert.

In order to make this a reality, I needed to do a few things

* Add tests to `config`. Since I was going to touch it, I thought that more tests were going to be required
* Upgrade go to 1.15 due to a go-jira dependency: https://github.com/golang-jwt/jwt
* Upgrade go-jira to 1.15.1 version to support PAT
* Add PAT support
* Test PAT changes in `config` package

Since some of the changes are related/needed but not exactly what the PR says, I've separated the above actions in different commits. Please let me know if you want them in separate PRs.

Lastly, I haven't written lots of Go code in my life, so apologies in advance for any obvious mistakes or not proper practices this code may have 😅

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -158,11 +159,18 @@ func (c *Config) UnmarshalYAML(unmarshal func(interface{}) error) error {
 	// We want to set c to the defaults and then overwrite it with the input.
 	// To make unmarshal fill the plain data struct rather than calling UnmarshalYAML
 	// again, we have to hide it using a type indirection.
+
+	// TODO: Handle the empty defaults case
```
Comment (bwplotka) [inline pkg/config/config.go:163]: What do you mean?

File: cmd/jiralert/main.go [hunk 2]
```diff
@@ -102,11 +102,24 @@ func main() {
...
+			}
+			client, err = jira.NewClient(tp.Client(), conf.APIURL)
+		} else if conf.PersonalAccessToken != "" {
+			tp := jira.PATAuthTransport{
+				Token: string(conf.PersonalAccessToken),
+			}
+			client, err = jira.NewClient(tp.Client(), conf.APIURL)
+		} else {
+			// Config should not allow us to get here, but just in case...
+			errorHandler(w, http.StatusInternalServerError, fmt.Errorf("missing JIRA auth config"), conf.Name, &data, logger)
```
Comment (bwplotka) [inline cmd/jiralert/main.go:120]: I would consider having one place of checking config. Either here or in unmarshal, not mixed?

File: pkg/config/config.go [hunk 3]
```diff
@@ -178,17 +186,20 @@ func (c *Config) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
-			rc.User = c.Defaults.User
+
+		if rc.User != "" && rc.Password != "" && rc.PersonalAccessToken != "" {
+			return fmt.Errorf("bad auth config in receiver %q: user/password and PAT authentication are mutually exclusive", rc.Name)
 		}
-		if rc.Password == "" {
-			if c.Defaults.Password == "" {
-				return fmt.Errorf("missing password in receiver %q", rc.Name)
+
+		if (rc.User == "" || rc.Password == "") && rc.PersonalAccessToken == "" {
```
Comment (bwplotka) [inline pkg/config/config.go:194]: That is weird - so we replace both password AND user if e.g user is set but password is not? Sounds wrong 🤔

File: pkg/config/config.go [hunk 4]
```diff
@@ -178,17 +186,20 @@ func (c *Config) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+			if c.Defaults.User != "" && c.Defaults.Password != "" {
+				rc.User = c.Defaults.User
+				rc.Password = c.Defaults.Password
+			} else if c.Defaults.PersonalAccessToken != "" {
```
Comment (bwplotka) [inline pkg/config/config.go:198]: hmm, why not setting those defaults one by one, without caring about dependencies from each?

File: pkg/config/config_test.go [hunk 5]
```diff
@@ -88,5 +91,308 @@ func TestLoadFile(t *testing.T) {
...
+	Summary             string `yaml:"summary,omitempty"`
+	ReopenState         string `yaml:"reopen_state,omitempty"`
+	ReopenDuration      string `yaml:"reopen_duration,omitempty"`
+
+	Priority          string `yaml:"priority,omitempty"`
+	Description       string `yaml:"description,omitempty"`
+	WontFixResolution string `yaml:"wont_fix_resolution,omitempty"`
+	AddGroupLabels    bool   `yaml:"add_group_labels,omitempty"`
+
+	// TODO(rporres): Add support for these
```
Comment (bwplotka) [inline pkg/config/config_test.go:114]:
```suggestion
	// TODO(rporres): Add support for these.
```

File: pkg/config/config_test.go [hunk 6]
```diff
@@ -88,5 +91,308 @@ func TestLoadFile(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, testConf, string(content))
 
-	// TODO(bwplotka): Add proper test cases on config struct.
+}
+
+// A test version of the ReceiverConfig struct to create test yaml fixtures
```
Comment (bwplotka) [inline pkg/config/config_test.go:96]:
```suggestion
// A test version of the ReceiverConfig struct to create test yaml fixtures.
```

File: pkg/config/config_test.go [hunk 7]
```diff
@@ -88,5 +91,308 @@ func TestLoadFile(t *testing.T) {
...
+	Description       string `yaml:"description,omitempty"`
+	WontFixResolution string `yaml:"wont_fix_resolution,omitempty"`
+	AddGroupLabels    bool   `yaml:"add_group_labels,omitempty"`
+
+	// TODO(rporres): Add support for these
+	// Fields            map[string]interface{} `yaml:"fields,omitempty"`
+	// Components        []string               `yaml:"components,omitempty"`
+}
+
+// A test version of the Config struct to create test yaml fixtures
```
Comment (bwplotka) [inline pkg/config/config_test.go:119]:
```suggestion
// A test version of the Config struct to create test yaml fixtures.
```

And below

File: pkg/config/config_test.go [hunk 8]
```diff
@@ -88,5 +91,308 @@ func TestLoadFile(t *testing.T) {
...
+		defaultsConfig := newReceiverTestConfig(test.receiverTestConfigMandatoryFields, []string{})
+		config := testConfig{
+			Defaults:  defaultsConfig,
+			Receivers: []*receiverTestConfig{minimalReceiverTestConfig},
+			Template:  "jiralert.tmpl",
+		}
+
+		configErrorTestRunner(t, config, test.errorMessage)
+	}
+}
```
Comment (bwplotka) [inline pkg/config/config_test.go:217]: This is great! 

But there is one pattern we could reuse here to simplify code (and have consistency).

```suggestion
	mandatory := mandatoryReceiverFields()

	minimalReceiverTestConfig := newReceiverTestConfig([]string{"Name"}, []string{})
	for _, test := range []struct {
		receiverTestConfigMandatoryFields []string
		errorMessage                      string
	}{
	      {
	      	      //  Missing user.
			removeFromStrSlice(mandatory, "User"),
			`missing authentication in receiver "Name"`,
		},
		{
		        // Missing password.
			removeFromStrSlice(mandatory, "Password"),
			`missing authentication in receiver "Name"`,
		},
		{
		       // Specifying user/password and PAT auth.
			append(mandatory, "PersonalAccessToken"),
			"bad auth config in defaults section: user/password and PAT authentication are mutually exclusive",
		},
	} {
		defaultsConfig := newReceiverTestConfig(test.receiverTestConfigMandatoryFields, []string{})
		config := testConfig{
			Defaults:  defaultsConfig,
			Receivers: []*receiverTestConfig{minimalReceiverTestConfig},
			Template:  "jiralert.tmpl",
		}

		configErrorTestRunner(t, config, test.errorMessage)
	}
}
```

**Review by bwplotka — COMMENTED** [review]
This is great! 

And for all tests - amazing work.

Couple of nits:

* Consider moving all config checks to one place
* Please add trailing dot to all comments (we follow Full sentence English)
* Table tests  are great but consider consistent, smaller implementation wit anonymous struct and in place test cases

File: pkg/config/config.go (hunk 1 above)
Comment (rporres) [reply pkg/config/config.go:163]: If there is no `defaults` section, jiralert panics

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x13c40b5]

goroutine 1 [running]:
gopkg.in/yaml%2ev2.handleErr(0xc000157d58)
	/Users/rporresm/go/1.15.15/pkg/mod/gopkg.in/yaml.v2@v2.3.0/yaml.go:249 +0x85
panic(0x15b0f60, 0x1a250c0)
	/Users/rporresm/.goenv/versions/1.15.15/src/runtime/panic.go:969 +0x1b9
github.com/prometheus-community/jiralert/pkg/config.(*Config).UnmarshalYAML(0xc00010f500, 0xc00000d860, 0x0, 0x94b6b18)
	/Users/rporresm/git/github.com/rporres/jiralert/pkg/config/config.go:170 +0x75
gopkg.in/yaml%2ev2.(*decoder).callUnmarshaler(0xc00002f080, 0xc000175ea0, 0x94b6b18, 0xc00010f500, 0xc00010f500)
	/Users/rporresm/go/1.15.15/pkg/mod/gopkg.in/yaml.v2@v2.3.0/decode.go:270 +0xa6
gopkg.in/yaml%2ev2.(*decoder).prepare(0xc00002f080, 0xc000175ea0, 0x1606360, 0xc00010f500, 0x199, 0x104e0a5, 0xc00000e140, 0x0, 0x0)
	/Users/rporresm/go/1.15.15/pkg/mod/gopkg.in/yaml.v2@v2.3.0/decode.go:313 +0x1d0
gopkg.in/yaml%2ev2.(*decoder).unmarshal(0xc00002f080, 0xc000175ea0, 0x1606360, 0xc00010f500, 0x199, 0x0)
	/Users/rporresm/go/1.15.15/pkg/mod/gopkg.in/yaml.v2@v2.3.0/decode.go:364 +0xe9
gopkg.in/yaml%2ev2.(*decoder).document(0xc00002f080, 0xc000175e30, 0x1606360, 0xc00010f500, 0x199, 0x13913a7)
	/Users/rporresm/go/1.15.15/pkg/mod/gopkg.in/yaml.v2@v2.3.0/decode.go:384 +0x7c
gopkg.in/yaml%2ev2.(*decoder).unmarshal(0xc00002f080, 0xc000175e30, 0x1606360, 0xc00010f500, 0x199, 0x199)
	/Users/rporresm/go/1.15.15/pkg/mod/gopkg.in/yaml.v2@v2.3.0/decode.go:360 +0x250
gopkg.in/yaml%2ev2.unmarshal(0xc000132800, 0x3f4, 0x400, 0x15c02c0, 0xc00010f500, 0x400, 0x0, 0x0)
	/Users/rporresm/go/1.15.15/pkg/mod/gopkg.in/yaml.v2@v2.3.0/yaml.go:148 +0x33d
gopkg.in/yaml%2ev2.Unmarshal(...)
	/Users/rporresm/go/1.15.15/pkg/mod/gopkg.in/yaml.v2@v2.3.0/yaml.go:81
github.com/prometheus-community/jiralert/pkg/config.Load(0xc000132400, 0x3f4, 0x3f4, 0xc000132400, 0x3f4)
	/Users/rporresm/git/github.com/rporres/jiralert/pkg/config/config.go:53 +0x99
github.com/prometheus-community/jiralert/pkg/config.LoadFile(0x1657b30, 0x13, 0x16fb140, 0xc000286e10, 0x0, 0x0, 0xc000286f00, 0x0, 0x0, 0x0)
	/Users/rporresm/git/github.com/rporres/jiralert/pkg/config/config.go:67 +0x214
main.main()
	/Users/rporresm/git/github.com/rporres/jiralert/cmd/jiralert/main.go:74 +0x2a5

File: cmd/jiralert/main.go (hunk 2 above)
Comment (rporres) [reply cmd/jiralert/main.go]: Fair enough. Fixed in ccb0fda

[the remark this answers is not in the document]
Comment (rporres) [reply cmd/jiralert/main.go]: This section was removed from ccb0fda as a result of https://github.com/prometheus-community/jiralert/pull/110#discussion_r824532687

File: pkg/config/config.go (hunk 3 above)
Comment (rporres) [reply pkg/config/config.go:194]: Hmmm, I was considering user & password as a block all the time, but it's true that being separate keys, they should be overridable separately. I will take a look to the logic again.

File: pkg/config/config.go (hunk 4 above)
Comment (rporres) [reply pkg/config/config.go:198]: Same thing as in https://github.com/prometheus-community/jiralert/pull/110#discussion_r825512228.

File: pkg/config/config.go (hunk 1 above)
Comment (rporres) [reply pkg/config/config.go]: Clarified comment in b7f35ac

File: pkg/config/config_test.go (hunk 8 above)
Comment (rporres) [reply pkg/config/config_test.go]: Fixed this everywhere in b2861cf

**Comment by hanzz** [comment]
Can we release new version once this is merged and tested? There is a move from basic auth to PAT in my company and we rely on jiralert for monitoring...

File: pkg/config/config.go (hunk 3 above)
Comment (rporres) [reply pkg/config/config.go:194]: Fixed in 107bc8f and added tests. User & Password can be overriden separately. The code has become a bit messy as there are a lot of passible scenarios to cover. I couldn't figure out something much cleaner 😓

File: pkg/config/config.go (hunk 4 above)
Comment (rporres) [reply pkg/config/config.go:205]: resolving in favour of https://github.com/prometheus-community/jiralert/pull/110#discussion_r825512228

**Comment by rporres** [comment]
This should be ready for a new review, @bwplotka. I've added separate commits to address comments separately. Happy to rearrange commits whenever this is ready to go.

**Review by bwplotka — APPROVED** [review]
Thanks, amazing job!

**Pull request merged**